### PR TITLE
Fix visibility check at definition for aspects

### PR DESCRIPTION
**Pull request by fmeum**
With `--incompatible_visibility_private_attributes_at_definition`, the visibility of aspects (and aspects on aspects) into their implicit dependencies is now checked relative to the aspect's, not the rule's definition.

This requires ensuring the precedence of aspect on aspect attributes over those of the underlying aspect in `DependencyResolutionHelper`, which was hinted at by a comment but not realized.

### CODE REVIEW

**Comment by fmeum** [comment]
@mai93 @comius was right in https://github.com/bazelbuild/bazel/pull/19141#issuecomment-1708299889, the aspect case of `--incompatible_visibility_private_attributes_at_definition` wasn't implemented correctly. 

I think that visibility for an implicit attribute of an aspect should be checked relative to that aspect's definition. I have implemented this in this PR and added test cases.

The new tests uncovered the interesting edge case of which attribute wins if two aspects define attributes with the same name. In the spirit of @mai93's commit https://github.com/bazelbuild/bazel/commit/93ad5890883b7591eef23ae3c3334a245c2aa6ef, I would argue that the outer aspect should see its own attribute or it might risk being broken by being applied to an aspect with an attribute of an incompatible type. `aspectWithExtraAttributeDependsOnNotApplicable_usesAttributeFromDependentAspect` codifies the opposite behavior though.

Is flipping that behavior around something we could do? If not, could we do that for implicit attributes only? Picking up an implicit attribute from a lower aspect instead of the own one seems just wrong.

**Comment by mai93** [comment]
Thanks @fmeum! I also think the main aspect should get its dependencies from its own attributes not from its underlying aspects. I need to run internal tests for this part to verify that there is nothing depending on the old behaviour that needs to be updated first.

**Comment by fmeum** [comment]
@mai93 Thanks for the feedback. I will separate out the change to aspect ordering.

I identified another place in which this ordering would probably need to be changed, but it doesn't seem to be covered by tests: https://github.com/bazelbuild/bazel/pull/19449
In a previous version of that test PR, I replaced `putIfAbsent` with `put`, but all tests passed (except one that seemed unrelated). Do you happen to have an idea why this doesn't have an effect on the existing aspect order tests?

File: src/test/java/com/google/devtools/build/lib/analysis/AspectTest.java [hunk 1]
```diff
@@ -1055,7 +1055,7 @@ public void aspectWithExtraAttributeDependsOnNotApplicable_usesAttributeFromDepe
     ExtraAttributeAspect.Provider provider =
         getAspectByName(analysisResult.getAspectsMap(), aspectApplies.getName())
             .getProvider(ExtraAttributeAspect.Provider.class);
-    assertThat(provider.label()).isEqualTo("//extra:extra2");
+    assertThat(provider.label()).isEqualTo("//extra:extra");
```
[the remark this answers is not in the document]
Comment (fmeum) [reply src/test/java/com/google/devtools/build/lib/analysis/AspectTest.java:1058]: Extracted into #19449

**Comment by fmeum** [comment]
@mai93 I extracted the aspect-on-aspect fixes into #19449 and stacked the visibility fixes onto that PR.

**Comment by fmeum** [comment]
@mai93 I set the test blocked on #19449 to ignored as you suggested in https://github.com/bazelbuild/bazel/pull/19449#issuecomment-1716979331.

File: src/main/java/com/google/devtools/build/lib/analysis/CommonPrerequisiteValidator.java [hunk 2]
```diff
@@ -105,21 +108,45 @@ private void validateDirectPrerequisiteVisibility(
...
+      }
       if (!RuleContext.isVisible(implicitDefinition, prerequisite.getConfiguredTarget())) {
         handleVisibilityConflict(context, prerequisite, implicitDefinition);
       }
     }
   }
 
+  private static boolean forStarlarkRuleOrAspect(RuleContext.Builder context) {
+    Aspect currentAspect = context.getMainAspect();
+    if (currentAspect != null && currentAspect.getAspectClass() instanceof StarlarkAspectClass) {
```
Comment (mai93) [inline src/main/java/com/google/devtools/build/lib/analysis/CommonPrerequisiteValidator.java:144]: I missed this in the first review but if the aspect is native, this needs to return `False` and not fallback to the rule type.

Comment (fmeum) [reply src/main/java/com/google/devtools/build/lib/analysis/CommonPrerequisiteValidator.java]: Good catch, fixed. Also renamed the local variable to `mainAspect`.

**Review by mai93 — APPROVED** [review]

File: src/main/java/com/google/devtools/build/lib/analysis/CommonPrerequisiteValidator.java [hunk 3]
```diff
@@ -105,21 +108,46 @@ private void validateDirectPrerequisiteVisibility(
     if (!toolCheckAtDefinition
         || !attribute.isImplicit()
         || attribute.getName().equals(RuleClass.CONFIG_SETTING_DEPS_ATTRIBUTE)
-        || rule.getRuleClassObject().getRuleDefinitionEnvironmentLabel() == null) {
+        || !forStarlarkRuleOrAspect(context)) {
```
Comment (comius) [inline src/main/java/com/google/devtools/build/lib/analysis/CommonPrerequisiteValidator.java:111]: Readability: rename `forStarlarkRuleOrAspect` to `inStarlarkRuleOrAspect` and move it into the RuleContext.Builder.

File: src/main/java/com/google/devtools/build/lib/analysis/RuleContext.java [hunk 4]
```diff
@@ -2091,6 +2091,11 @@ public boolean isVisible(TransitiveInfoCollection prerequisite) {
       return RuleContext.isVisible(target.getAssociatedRule(), prerequisite);
     }
 
+    @Nullable
+    Aspect getMainAspect() {
```
Comment (comius) [inline src/main/java/com/google/devtools/build/lib/analysis/RuleContext.java:2095]: Readability: Can you reuse the same implementation as `getMainAspect` in the RuleContext? That would be still some duplication, but at least not a different one.

File: src/main/java/com/google/devtools/build/lib/analysis/CommonPrerequisiteValidator.java [hunk 5]
```diff
@@ -105,21 +108,46 @@ private void validateDirectPrerequisiteVisibility(
...
+        implicitDefinition =
+            checkNotNull(rule.getRuleClassObject().getRuleDefinitionEnvironmentLabel());
+      }
       if (!RuleContext.isVisible(implicitDefinition, prerequisite.getConfiguredTarget())) {
         handleVisibilityConflict(context, prerequisite, implicitDefinition);
       }
     }
   }
 
+  private static boolean forStarlarkRuleOrAspect(RuleContext.Builder context) {
```
Comment (comius) [inline src/main/java/com/google/devtools/build/lib/analysis/CommonPrerequisiteValidator.java:142]: Readability: move `forStarlarkRuleOrAspect` into the RuleContext.Builder.
Rename `forStarlarkRuleOrAspect` to `isStarlarkRuleOrAspect` (I see there is preexsiting use of `for`, but I think is would make more sense)

File: src/main/java/com/google/devtools/build/lib/analysis/CommonPrerequisiteValidator.java [hunk 6]
```diff
@@ -105,21 +108,46 @@ private void validateDirectPrerequisiteVisibility(
...
+      // from the location of the definition that declares the attribute. Only perform this check
+      // for the current aspect.
+      Label implicitDefinition;
+      Aspect mainAspect = context.getMainAspect();
+      if (mainAspect != null) {
+        if (!mainAspect.getDefinition().getAttributes().containsKey(attrName)) {
+          // Only verify visibility of implicit dependencies of the current aspect. Implicit
+          // dependencies of other aspects as well as the rule itself are checked when they are
+          // evaluated.
+          return;
```
Comment (comius) [inline src/main/java/com/google/devtools/build/lib/analysis/CommonPrerequisiteValidator.java:127]: Readability: The `return` here breaks the encapsulation of the else block. You can't add any code after the block and assume it's correct.

Possible solution is to invert `if (!mainAspect.getDefinition().getAttributes().containsKey(attrName)) {`
and move the block `if (!RuleContext.isVisible...` inside after setting `implicitDefinition`.

File: src/test/java/com/google/devtools/build/lib/analysis/VisibilityTest.java [hunk 7]
```diff
@@ -202,6 +203,407 @@ public void testConfigSettingVisibilityAlwaysCheckedAtUse() throws Exception {
     assertThat(hasErrors(getConfiguredTarget("//:my_target"))).isFalse();
   }
 
+  @Test
+  public void testAspectImplicitDependencyCheckedAtDefinition_visible() throws Exception {
```
Comment (comius) [inline src/test/java/com/google/devtools/build/lib/analysis/VisibilityTest.java:207]: Readability: All the tests should be structured as arrange, act, assert, with an empty line between the blocks.

Remove empty lines between `scratch` calls everywhere. And remove empty lines before `useConfiguration` - keep it together with `scratch`.

File: src/test/java/com/google/devtools/build/lib/analysis/VisibilityTest.java [hunk 8]
```diff
@@ -202,6 +203,407 @@ public void testConfigSettingVisibilityAlwaysCheckedAtUse() throws Exception {
...
+        "sh_binary(",
+        "  name = 'rule_tool',",
+        "  srcs = ['a.sh'],",
+        "  visibility = ['//rule:__pkg__'],",
+        ")");
+
+    useConfiguration("--incompatible_visibility_private_attributes_at_definition");
+
+    update("//foo:target_with_aspects");
+    assertThat(hasErrors(getConfiguredTarget("//foo:target_with_aspects"))).isFalse();
```
Comment (comius) [inline src/test/java/com/google/devtools/build/lib/analysis/VisibilityTest.java:277]: This should be:
```
scratch
useConfiguration // arrange

update // act

assertThat // assert
```

Same in `testAspectImplicitDependencyCheckedAtDefinition_visibleWithNameCollision`

**Review by comius — CHANGES_REQUESTED** [review]

File: src/main/java/com/google/devtools/build/lib/analysis/CommonPrerequisiteValidator.java (hunk 6 above)
Comment (fmeum) [reply src/main/java/com/google/devtools/build/lib/analysis/CommonPrerequisiteValidator.java]: I wasn't entirely sure what you meant with "move the block ... inside". I left it outside without duplicating it, please let me know what you think.

File: src/test/java/com/google/devtools/build/lib/analysis/VisibilityTest.java (hunk 7 above)
Comment (fmeum) [reply src/test/java/com/google/devtools/build/lib/analysis/VisibilityTest.java:207]: I tried to adapt all tests to that convention, but wasn't sure how to handle `assertThrows`, which is both "act" and "assert".

Comment (comius) [reply src/test/java/com/google/devtools/build/lib/analysis/VisibilityTest.java:207]: Yeah, we don't have strict guidelines for that. One empty line works. Thanks.

**Review by comius — APPROVED** [review]

**Comment by comius** [comment]
This is ready for reimport. @iancha1992
